### PR TITLE
[react] Puzzle tag API and UI

### PR DESCRIPTION
**Pull request by rgossiaux**
Add in the functionality to add/remove existing tags to a puzzle and to create new tags.

It looks a little shitty and the UX is a little shitty too but for the moment I'm just worried about the functionality.

### CODE REVIEW

**Review by akirabaruah — CHANGES_REQUESTED** [review]
This is great! If it's not too tedious, we should probably add some basic unit tests for the API. There are some good error cases that we probably don't want to regress on.

File: api/urls.py [hunk 1]
```diff
@@ -16,8 +16,29 @@
...
+tag_list = views.PuzzleTagViewSet.as_view(
+    {
+        "post": "create",
+    }
+)
+
+tag_detail = views.PuzzleTagViewSet.as_view(
+    {
+        "get": "retrieve",
+        "delete": "destroy",
```
[the remark this answers is not in the document]
Comment (rgossiaux) [reply api/urls.py:28]: These names come from DRF: https://www.django-rest-framework.org/api-guide/viewsets/#viewset-actions

File: api/views.py [hunk 2]
```diff
@@ -108,3 +108,70 @@ def create(self, request, **kwargs):
...
+        with transaction.atomic():
+            hunt = get_object_or_404(Hunt, pk=self.kwargs["hunt_id"])
+            puzzle = get_object_or_404(Puzzle, pk=self.kwargs["puzzle_id"])
+            serializer = self.get_serializer(data=request.data, context={"hunt": hunt})
+            serializer.is_valid(raise_exception=True)
+            (tag_name, tag_color) = (
+                serializer.validated_data["name"],
+                serializer.validated_data["color"],
+            )
+            (tag, _) = PuzzleTag.objects.update_or_create(
```
[the remark this answers is not in the document]
Comment (rgossiaux) [reply api/views.py]: Sure (this was actually copied from puzzles/views.py)

**Comment by rgossiaux** [comment]
Yeah, I was planning on adding tests for all this/porting our existing tests over after finishing the feature work / before "launching", though there's an argument that it's better to do it while things are fresher. It's one of the checklist items in #251

**Comment by akirabaruah** [comment]
Okay yeah that sounds fine, as long we're planning to add tests relatively
soon. +1 that adding tests sooner is nice, but I can understand that we
want to get the minimum features up ASAP. Thanks for clarifying the plan.

File: api/views.py [hunk 3]
```diff
@@ -108,3 +108,70 @@ def create(self, request, **kwargs):
...
+                    return Response(
+                        {
+                            "detail": '"Unable to assign metapuzzle since doing so would introduce a meta-cycle."'
+                        },
+                        status=400,
+                    )
+                # the post m2m hook will add tag
+                puzzle.metas.add(metapuzzle)
+            else:
+                puzzle.tags.add(tag)
```
Comment (asdfryan) [inline api/views.py:175]: Do we need to call puzzle.save()?

File: hunts/src/TagCell.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import React from "react";
+import Badge from "react-bootstrap/Badge";
+import { useDispatch, useSelector } from "react-redux";
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
+import { faPlus } from "@fortawesome/free-solid-svg-icons";
+import { deletePuzzleTag, fetchPuzzles } from "./puzzlesSlice";
```
Comment (asdfryan) [inline hunts/src/TagCell.js:6]: These imports seem no longer needed?

File: api/views.py (hunk 3 above)
Comment (rgossiaux) [reply api/views.py:175]: No, `.add` saves the relationship in the DB: https://docs.djangoproject.com/en/3.1/topics/db/examples/many_to_many/

File: api/views.py [hunk 5]
```diff
@@ -6,8 +6,8 @@
 from rest_framework import viewsets
 
 from hunts.models import Hunt
-from puzzles.models import Puzzle, PuzzleModelError
-from .serializers import HuntSerializer, PuzzleSerializer
+from puzzles.models import Puzzle, PuzzleModelError, PuzzleTag, is_ancestor
```
Comment (akirabaruah) [inline api/views.py:9]: Probably beyond the scope of this PR, but shouldn't we make `is_ancestor` a method of Puzzle? Seems like `puzzle_object.is_ancestor_of(other_puzzle_object)` would make more sense.

File: api/views.py [hunk 6]
```diff
@@ -108,3 +108,70 @@ def create(self, request, **kwargs):
...
+                return Response(
+                    {
+                        "detail": "You cannot remove a meta's tag (%s) from itself"
+                        % tag.name
+                    },
+                    status=400,
+                )
+            if tag.is_meta:
+                meta = Puzzle.objects.get(name=tag.name, hunt=puzzle.hunt)
+                # the post m2m hook will remove tag
```
Comment (akirabaruah) [inline api/views.py:136]: Have we considered creating m2m hooks on the PuzzleTag object that checks `is_meta` and makes these updates accordingly? That way, we could probably just call `puzzle.tags.remove(tag)` here and `puzzle.tag.add(tag)` below and avoid having these conditionals. It seems a bit strange to use the `metas` field in some cases and `tags` field in other cases, when we're really creating/deleting tag objects under the hood.

**Review by akirabaruah — APPROVED** [review]
LGTM overall, modulo one question about possibly simplifying the Tag creation/deletion logic 👍

File: api/views.py (hunk 5 above)
Comment (rgossiaux) [reply api/views.py:9]: I agree, good suggestion.

File: api/views.py (hunk 6 above)
Comment (rgossiaux) [reply api/views.py:136]: Yeah, that makes sense to me (this was just the existing structure copied over to this file), especially if we're always going to be going in the tag => meta direction.

**Pull request merged**